pkg/istio-agent/health/health_check_test.go

```diff
@@ -15,127 +15,150 @@
+					tcpFinishedEvents.Wait()
+				}
+			}
+		}()
 
-	// wait for go-ahead for state change
-	go func() {
-		for i := 0; i < len(tcpHealthStatuses); i++ {
-			if tcpHealthStatuses[i] {
-				// open port until we get confirmation that
-				srv, _ := net.Listen("tcp", "localhost:5991")
-				// add to delta and wait for event received
-				tcpFinishedEvents.Add(1)
-				tcpFinishedEvents.Wait()
-				srv.Close()
-			} else {
-				// dont listen or anything, just wait for event to finish to
-				// start next iteration
-				tcpFinishedEvents.Add(1)
-				tcpFinishedEvents.Wait()
+		eventNum := atomic.NewInt32(0)
+		go tcpHealthChecker.PerformApplicationHealthCheck(func(event *ProbeEvent) {
+			tcpFinishedEvents.Done()
+			if event.Healthy != expectedTCPEvents[eventNum.Load()].Healthy {
+				t.Errorf("%s: got event healthy: %v at idx %v when expected healthy: %v", "tcp", event.Healthy, eventNum.Load(), expectedTCPEvents[eventNum.Load()].Healthy)
+			}
+			eventNum.Inc()
+		}, quitChan)
+		retry.UntilSuccessOrFail(t, func() error {
+			if int(eventNum.Load()) != len(expectedTCPEvents) {
+				return fmt.Errorf("waiting for %v events", len(expectedTCPEvents)-int(eventNum.Load()))
 			}
+			return nil
+		}, retry.Delay(time.Millisecond*10), retry.Timeout(time.Second))
+		close(quitChan)
+	})
+	t.Run("http", func(t *testing.T) {
+		httpPath := "/test/health/check"
+		httpHealthChecker := NewWorkloadHealthChecker(&v1alpha3.ReadinessProbe{
+			InitialDelaySeconds: 0,
+			TimeoutSeconds:      1,
+			PeriodSeconds:       1,
+			SuccessThreshold:    1,
+			FailureThreshold:    1,
+			HealthCheckMethod: &v1alpha3.ReadinessProbe_HttpGet{
+				HttpGet: &v1alpha3.HTTPHealthCheckConfig{
+					Path:   httpPath,
+					Port:   65112,
+					Scheme: "http",
+					Host:   "127.0.0.1",
+				},
+			},
+		})
+		// Speed up tests
+		httpHealthChecker.config.CheckFrequency = time.Millisecond
+		quitChan := make(chan struct{})
+		expectedHTTPEvents := [4]*ProbeEvent{
+			{Healthy: true},
+			{Healthy: false},
+			{Healthy: true},
+			{Healthy: false},
 		}
-	}()
-
-	go tcpHealthChecker.PerformApplicationHealthCheck(tcpHealthChan, quitChan)
+		httpHealthStatuses := [4]bool{true, false, true, false}
 
-	for c := 0; c < len(expectedTCPEvents); c++ {
-		probeEvent := <-tcpHealthChan
-		tcpFinishedEvents.Done()
-		if probeEvent.Healthy != expectedTCPEvents[c].Healthy {
-			t.Errorf("%s: got event healthy: %v at idx %v when expected healthy: %v", "tcp", probeEvent.Healthy, c, expectedTCPEvents[c].Healthy)
+		mockListener, err := net.Listen("tcp", "127.0.0.1:65112")
+		if err != nil {
+			t.Errorf("unable to start mock listener: %v", err)
 		}
-	}
-	quitChan <- struct{}{}
+		httpServerEventCount := 0
+		srv := httptest.NewUnstartedServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
+			if httpServerEventCount < len(httpHealthStatuses) && httpHealthStatuses[httpServerEventCount] {
+				writer.WriteHeader(200)
+				writer.Write([]byte("foobar"))
```

Comment:
is this necessary? are we expecting a body?

Reply:
Its not necessary, it was just already there. I think its probably best to test with and without body if we want to be complete, although right now we don't.